File: fizzbuzz.cpp
#include <iostream>
#include <vector>
#include <string>
#include <variant>

void FizzBuzz(const std::vector<std::variant<int, std::string>>& rules, int upperLimit, int currentNumber, int ruleIndex, std::string output) {
    if (ruleIndex < rules.size()) {
        if (currentNumber % std::get<int>(rules[ruleIndex]) == 0) {
            output += std::get<std::string>(rules[ruleIndex + 1]);
        }
        FizzBuzz(rules, upperLimit, currentNumber, ruleIndex + 2, output);
    } else {
        if (currentNumber <= upperLimit) {
            if (!output.empty()) {
                std::cout << output << std::endl;
            } else {
                std::cout << currentNumber << std::endl;
            }
            FizzBuzz(rules, upperLimit, currentNumber + 1, 0, "");
        }
    }
}

int main() {
    std::vector<std::variant<int, std::string>> rules = { 3, "fizz", 5, "buzz" };
    int upperLimit = 100;
    int currentNumber = 0;
    int ruleIndex = 0;
    std::string output = "";

    FizzBuzz(rules, upperLimit, currentNumber, ruleIndex, output);

    return 0;
}

File: fizzbuzz.py
def FizzBuzz(rules, n, curr_num, rule_idx, out):
    if rule_idx < len(rules):
        if curr_num % rules[rule_idx] == 0:
            out += rules[rule_idx + 1]
        return FizzBuzz(rules, n, curr_num, rule_idx + 2, out) or (not rule_idx and curr_num < n and FizzBuzz(rules, n, curr_num + 1, 0, '')) or 0
    else:
        print(out or curr_num)

FizzBuzz([3, 'fizz', 5, 'buzz'], 100, 0, 0, '')

